libtextsecure/message_receiver.js

```diff
@@ -661,19 +663,17 @@ MessageReceiver.prototype.extend({
         var sessionCipher = new libsignal.SessionCipher(textsecure.storage.protocol, address, options);
         console.log('retrying prekey whisper message');
         return this.decryptPreKeyWhisperMessage(ciphertext, sessionCipher, address).then(function(plaintext) {
-            var finalMessage = textsecure.protobuf.DataMessage.decode(plaintext);
-
-            var p = Promise.resolve();
-            if ((finalMessage.flags & textsecure.protobuf.DataMessage.Flags.END_SESSION)
-                    == textsecure.protobuf.DataMessage.Flags.END_SESSION &&
-                    finalMessage.sync !== null) {
-                    var number = address.getName();
-                    p = this.handleEndSession(number);
-            }
-
-            return p.then(function() {
-                return this.processDecrypted(finalMessage);
-            }.bind(this));
+            var now = Date.now();
+            var envelope = {
+                source: number,
+                sourceDevice: device,
+                timestamp: {
+                    toNumber: function() {
+                        return now;
```

Comment:
Right now the retry mechanism only includes the source address and the serialized content of the message. I'll look into passing more data along - namely, the original receipt time.